flink/src/test/java/org/apache/iceberg/flink/TestFlinkFilters.java

```diff
@@ -0,0 +1,361 @@
+      .field("field7", DataTypes.BINARY(2))
+      .field("field8", DataTypes.DECIMAL(10, 2))
+      .field("field9", DataTypes.DATE())
+      .field("field10", DataTypes.TIME())
+      .field("field11", DataTypes.TIMESTAMP())
+      .field("field12", DataTypes.TIMESTAMP_WITH_LOCAL_TIME_ZONE())
+      .build();
+
+  // A map list of fields and values used to verify the conversion of flink expression to iceberg expression
+  private static final List<Tuple2<String, Object>> FIELD_VALUE_LIST = ImmutableList.of(
+      Tuple2.of("field1", 1),
+      Tuple2.of("field2", 2L),
+      Tuple2.of("field3", 3F),
+      Tuple2.of("field4", 4D),
+      Tuple2.of("field5", "iceberg"),
+      Tuple2.of("field6", true),
+      Tuple2.of("field7", new byte[] {'a', 'b'}),
+      Tuple2.of("field8", BigDecimal.valueOf(10)),
+      Tuple2.of("field9", DateTimeUtil.daysFromDate(LocalDate.now())),
+      Tuple2.of("field10", DateTimeUtil.microsFromTime(LocalTime.now())),
+      Tuple2.of("field11", DateTimeUtil.microsFromTimestamp(LocalDateTime.now())),
+      Tuple2.of("field12", DateTimeUtil.microsFromInstant(Instant.now()))
+  );
+
+  @Test
+  public void testFlinkDataTypeEqual() {
+    matchLiteral("field1", 1, 1);
+    matchLiteral("field2", 10L, 10L);
+    matchLiteral("field3", 1.2F, 1.2F);
+    matchLiteral("field4", 3.4D, 3.4D);
+    matchLiteral("field5", "abcd", "abcd");
+    matchLiteral("field6", true, true);
+    matchLiteral("field7", new byte[] {'a', 'b'}, ByteBuffer.wrap(new byte[] {'a', 'b'}));
+    matchLiteral("field8", BigDecimal.valueOf(10), BigDecimal.valueOf(10));
+
+    LocalDate date = LocalDate.parse("2020-12-23");
+    matchLiteral("field9", date, DateTimeUtil.daysFromDate(date));
+
+    LocalTime time = LocalTime.parse("12:13:14");
+    matchLiteral("field10", time, DateTimeUtil.microsFromTime(time));
+
+    LocalDateTime dateTime = LocalDateTime.parse("2020-12-23T12:13:14");
+    matchLiteral("field11", dateTime, DateTimeUtil.microsFromTimestamp(dateTime));
+
+    Instant instant = Instant.parse("2020-12-23T12:13:14.00Z");
+    matchLiteral("field12", instant, DateTimeUtil.microsFromInstant(instant));
+  }
+
+  @Test
+  public void testEquals() {
+    for (Tuple2<String, Object> tuple2 : FIELD_VALUE_LIST) {
+      UnboundPredicate<?> expected = org.apache.iceberg.expressions.Expressions.equal(tuple2.f0, tuple2.f1);
+
+      Optional<org.apache.iceberg.expressions.Expression> actual =
+          FlinkFilters.convert(resolve(Expressions.$(tuple2.f0).isEqual(Expressions.lit(tuple2.f1))));
+      Assert.assertTrue(actual.isPresent());
+      assertPredicatesMatch(expected, (UnboundPredicate<?>) actual.get());
+
+      Optional<org.apache.iceberg.expressions.Expression> actual1 =
+          FlinkFilters.convert(resolve(Expressions.lit(tuple2.f1).isEqual(Expressions.$(tuple2.f0))));
+      Assert.assertTrue(actual1.isPresent());
+      assertPredicatesMatch(expected, (UnboundPredicate<?>) actual1.get());
```

Comment:
It would be better to check the cast instead of failing tests with `ClassCastException` if the wrong class is returned. Could you update `assertPredicatesMatch` to accept `actual` as an `Expression` and add an `instanceof` assertion in that method?